src/coprocessor/codec/mysql/duration.rs

```diff
@@ -32,10 +42,152 @@ fn check_dur(dur: &StdDuration) -> Result<()> {
     Ok(())
 }
 
-fn tm_to_secs(t: Tm) -> u64 {
-    t.tm_hour as u64 * SECS_PER_HOUR + t.tm_min as u64 * SECS_PER_MINUTE + t.tm_sec as u64
+#[inline]
+fn check_hour(hour: u32) -> Result<u32> {
+    if hour > MAX_HOURS {
+        Err(invalid_type!(
+            "invalid hour value: {} larger than {}",
+            hour,
+            MAX_HOURS
+        ))
+    } else {
+        Ok(hour)
+    }
+}
+
+#[inline]
+fn check_minute(minute: u32) -> Result<u32> {
+    if minute > MAX_MINUTES {
+        Err(invalid_type!(
+            "invalid minute value: {} larger than {}",
+            minute,
+            MAX_MINUTES
+        ))
+    } else {
+        Ok(minute)
+    }
+}
+
+#[inline]
+fn check_second(second: u32) -> Result<u32> {
+    if second > MAX_SECONDS {
+        Err(invalid_type!(
+            "invalid second value: {} larger than {}",
+            second,
+            MAX_SECONDS
+        ))
+    } else {
+        Ok(second)
+    }
+}
+
+fn buf_to_int(buf: &[u8]) -> u32 {
+    buf.iter().fold(0, |acc, c| acc * 10 + u32::from(c - b'0'))
 }
 
+// Functionality:
+// Extract a `u32` from a buffer which matches pattern: `\d+.*`
```

Comment:
In [`nom::number`]( https://docs.rs/nom/5.0.0-alpha2/nom/number/complete/index.html ), the functions are used to parse integer/float/hex from binary instead of from `[u8]` where its entry store a  digit. In [`nom::character`](https://docs.rs/nom/5.0.0-alpha2/nom/character/complete/index.html), the functions are used to recognize buffer in some patterns without paring it into integer/float/hex.